### PR TITLE
Model Architecture | Training | Metrics | Validation Plots

### PR DESCRIPTION
**Pull request by TairYerniyazov**

### CODE REVIEW

File: hadml/models/cgan/multihadron_event_gan.py [hunk 1]
```diff
@@ -18,52 +18,92 @@ def __init__(
...
-                    d[:, :4] * torch.rand(1).to(device),           # scaling kinematics (Gaussian noise)
-                    torch.nn.functional.one_hot(d_type,            # randomly shifted types
-                                                num_classes=n_types)            
-                ]) for d, d_type in zip(disc_input, distorted_types)
-            ])
-            # ==========================================================================================
-            return {"distorted_disc_input": distorted_disc_input, "disc_input": disc_input}
-        
+            noise = self._generate_noise(*gen_input.size()[:2])
+            gen_output = self.generator(noise.to(gen_input.device), gen_input).detach()
```
Comment (remilvus) [inline hadml/models/cgan/multihadron_event_gan.py:103]: Is there a reason why we are not just calling `self(gen_input)` here?

File: hadml/models/cgan/multihadron_event_gan.py [hunk 2]
```diff
@@ -18,52 +18,92 @@ def __init__(
...
+            return {"gen_output": gen_output, "disc_input": disc_input.detach()}
         elif self.trainer.state.stage == "sanity_check":
```
Comment (remilvus) [inline hadml/models/cgan/multihadron_event_gan.py:105]: Does the validation step break during the sanity check? If it doesn't, I would remove this if-else statement. Otherwise, this is good as a fix.

File: hadml/models/components/encoder_transformer.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class Generator(torch.nn.Module):
+    """ Generator implemented as a encoder-only transformer model """
+
+    def __init__(
+        self,
+        noise_dim=4,            # Arbitrary number (noise dimensionality)
+        cluster_kins_dim=8,     # Cluster four-momentum, two quark types, phi, theta
+        n_quarks=2,             # Number of quark types in cluster_kins_dim
```
Comment (remilvus) [inline hadml/models/components/encoder_transformer.py:11]: Number of quark**s** in cluster_kins_dim

File: hadml/models/components/encoder_transformer.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import torch, os, pickle
+
+
+class Generator(torch.nn.Module):
+    """ Generator implemented as a encoder-only transformer model """
+
+    def __init__(
+        self,
+        noise_dim=4,            # Arbitrary number (noise dimensionality)
+        cluster_kins_dim=8,     # Cluster four-momentum, two quark types, phi, theta
```
Comment (remilvus) [inline hadml/models/components/encoder_transformer.py:10]: Does it make sense to have quark types included in this kinematics variable? Wouldn't `cluster_kins_dim=6` be better?

File: hadml/models/components/encoder_transformer.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        n_heads=8,              # Encoder architecture hyperparameter
+        pid_map_filepath=None   # For getting information about the number of hadron most common IDs
+    ):
+        super().__init__()
+        self.hadron_kins_dim = hadron_kins_dim
+        with open(os.path.normpath(pid_map_filepath), "rb") as f:
+            n_hadron_types = len(pickle.load(f)) + 1
+        self.quark_type_embedding_layer = torch.nn.Embedding(quark_types, quark_embedding_dim)
+        self.input_embedding_layer = torch.nn.Linear(
+            noise_dim + cluster_kins_dim - n_quarks + n_quarks * quark_embedding_dim, embedding_dim
```
Comment (remilvus) [inline hadml/models/components/encoder_transformer.py:25]: Removing quark types from kinematics would simplify this line a bit. Assuming that removing it makes sense.

File: hadml/models/components/encoder_transformer.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        )
+        self.output_embedding_layer = torch.nn.Linear(embedding_dim, hadron_kins_dim + n_hadron_types)
+        encoder_layer = torch.nn.TransformerEncoderLayer(d_model=embedding_dim, nhead=n_heads, batch_first=True)
+        self.transformer_encoder = torch.nn.TransformerEncoder(encoder_layer, num_layers=6)
+
+    def forward(self, noise, cluster_kins):
+        quark_types = cluster_kins[:, :, 4:6].to(torch.int32)
+        quark_types = torch.where(quark_types < 0, torch.abs(quark_types), quark_types + 8)
```
Comment (remilvus) [inline hadml/models/components/encoder_transformer.py:33]: This mapping looks like something that should happen during data loading/prep

File: hadml/models/cgan/multihadron_event_gan.py [hunk 7]
```diff
@@ -18,52 +18,92 @@ def __init__(
...
+        self.val_disc_loss = MeanMetric()
+        self.current_gumbel_temp = 1.0
 
     def forward(self, clusters):
-        generator_input = conditional_cat(clusters, self._generate_noise(len(clusters)))
-        generated_hadrons = self.hparams.generator(generator_input)
+        noise = self._generate_noise(*clusters.size()[:2])
+        generated_hadrons = self.generator(noise.to(clusters.device), clusters)
+        hadron_kins_dim = self.generator.hadron_kins_dim
+        generated_hadrons[:, :, hadron_kins_dim:] = torch.nn.functional.gumbel_softmax(
```
Comment (remilvus) [inline hadml/models/cgan/multihadron_event_gan.py:40]: Can you make the `hard` argument configurable through the config file ([link](https://pytorch.org/docs/stable/generated/torch.nn.functional.gumbel_softmax.html#torch-nn-functional-gumbel-softmax))? It would be worthwhile to test whether it improves the performance (I'm expecting that it will)

File: hadml/models/components/encoder_transformer.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    def __init__(
+        self,
+        hadron_kins_dim=4,      # Hadron four-momentum
+        embedding_dim=32,       # Arbitrary number (but the same for the discriminator)
+        n_heads=8,              # Encoder architecture hyperparameter
+        pid_map_filepath=None   # For getting information about the number of hadron most common IDs
+    ):
+        super().__init__()
+        with open(os.path.normpath(pid_map_filepath), "rb") as f:
+            n_hadron_types = len(pickle.load(f)) + 1
```
Comment (remilvus) [inline hadml/models/components/encoder_transformer.py:57]: Are we always using all hadron types, with no way to limit `n_hadron_types` to e.g. 10? We should be able to limit this (during data loading) to make debugging easier / simplify the problem. Sorry, I think I missed it when we discussed data loading.

File: configs/model/multihadron_event_gan.yaml [hunk 9]
```diff
@@ -1,18 +1,25 @@
 _target_: hadml.models.cgan.multihadron_event_gan.MultiHadronEventGANModule
 noise_dim: 4
+loss_type: "ls"
```
Comment (remilvus) [inline configs/model/multihadron_event_gan.yaml:3]: As we are following [this paper](https://arxiv.org/pdf/1801.04406) for regularization,  we should also keep to the assumptions they make on the loss there (meaning using e.g. "bce"). See eq. (1) and assumption II.

File: hadml/datamodules/components/herwig_multihadron_parser.py [hunk 10]
```diff
@@ -33,6 +33,10 @@ def __init__(self,
             raw_file_list,           # List of raw data filenames.
             processed_filename,      # Processed data filename.
             pid_map_file,            # PID-to-MostCommonID map filename.
+            n_hadron_types=None,     # Number of hadron types to be considered (None for all types).
```
Comment (remilvus) [inline hadml/datamodules/components/herwig_multihadron_parser.py:36]: Before, we dropped PIDs that were "uncommon", but this approach also looks sensible. Let's discuss during the next meeting

Comment (TairYerniyazov) [reply hadml/datamodules/components/herwig_multihadron_parser.py:36]: In the case of multihadron cluster decays, dropping samples containing unmappable PIDs would probably result in the loss of too much training data.

File: configs/model/multihadron_event_gan.yaml (hunk 9 above)
Comment (TairYerniyazov) [reply configs/model/multihadron_event_gan.yaml:3]: I opted for the least squares approach from the last sweep, as it was part of the set of hyperparameters that led to the lowest Wasserstein distance, though the entire sweep was run on a small amount of data. I will try running it on BCE.

File: hadml/datamodules/components/herwig_multihadron_parser.py (hunk 10 above)
Comment (remilvus) [reply hadml/datamodules/components/herwig_multihadron_parser.py:36]: If we dropped entire samples (like before) then most likely yes, but if we only dropped individual hadrons everything should work well. But this "uncommon" approach is probably better for event modelling.

File: configs/model/multihadron_event_gan.yaml (hunk 9 above)
Comment (remilvus) [reply configs/model/multihadron_event_gan.yaml:3]: I would always use bce loss to make the hyperparameter sweep simpler (and also because it's more theoretically grounded). Also (I'm guessing here), if the Wasserstein distance is lower, but the model isn't actually converging, we shouldn't draw any strong conclusions from it.

File: hadml/datamodules/gan_datamodule.py [hunk 11]
```diff
@@ -385,15 +390,44 @@ def setup(self, stage: Optional[str] = None):
...
                 generator=torch.Generator().manual_seed(42),
             )
-
+            
+            # Computing values (using the training data) needed to standardise kinematics
+            if os.path.exists(self.training_stats_filename):
+                print("Found training data statistics:\n   ", self.training_stats_filename,
+                      '\n', '-'*70, sep='')
+            else: 
+                hadron_kinematics, cluster_kinematics = [], []
```
Comment (remilvus) [inline hadml/datamodules/gan_datamodule.py:408]: Shouldn't all/most of this logic be a part of `set_training_stats`? Or at least a separate method? This would make for a more concise code here

File: hadml/datamodules/components/herwig.py [hunk 12]
```diff
@@ -678,24 +681,43 @@ def __init__(self, n_had_types, clusters, hadrons_with_types, max_n_hadrons):
...
     def __getitem__(self, index):
         """ This method is mainly responsible for tokenisation and building sentences containing 
         "cluster/hadron padding tokens". It returns a pair of prepared sentences. """
+        
+        cluster = self.clusters[index]
+        hadrons = self.hadrons_with_types[index].kinematics
+        
+        # Standardising cluster kinematics
+        if self.cluster_stats is not None:
+            cluster[0] = cluster[0] - self.cluster_stats["energy_mean"]
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:696]: `cluster_stats`/`hadron_stats` could have been a dataclass (so we know what fields are available). As far as I see, they both even share the exact same keys

File: hadml/datamodules/components/herwig.py [hunk 13]
```diff
@@ -704,5 +726,21 @@ def __getitem__(self, index):
...
                                             for _ in range(self.max_n_hadrons - n_hadrons)])
             disc_input = torch.cat([disc_input, hadron_padding_tokens])
+        return gen_input.to(torch.float32), disc_input.to(torch.float32)
         
-        return gen_input, disc_input
\ No newline at end of file
+    def get_kinematics(self, index):
+        # Returning kinematics only for computing values needed for standardisation
+        return {"hadron_kin": self.hadrons_with_types[index].kinematics,
+                "cluster_kin": self.clusters[index][:4]}
```
Comment (remilvus) [inline hadml/datamodules/components/herwig.py:734]: `self.clusters[index]` could potentially use named fields, similarly to  `self.hadrons_with_types[index]`

File: hadml/models/cgan/multihadron_event_gan.py [hunk 14]
```diff
@@ -18,52 +19,142 @@ def __init__(
...
         super().__init__()
         self.save_hyperparameters(ignore=["generator", "discriminator"])
         self.generator = generator
         self.discriminator = discriminator
+        self.train_gen_loss = MeanMetric()
+        self.train_disc_loss = MeanMetric()
+        self.val_gen_loss = MeanMetric()
+        self.val_disc_loss = MeanMetric()
+        self.current_gumbel_temp = 1.0
+        self.val_swd_1 = MeanMetric()
```
Comment (remilvus) [inline hadml/models/cgan/multihadron_event_gan.py:37]: Please improve this naming before we forget which is which 🙏
e.g. `val_swd_token`, `val_swd_sentence`

File: hadml/models/cgan/multihadron_event_gan.py [hunk 15]
```diff
@@ -73,9 +164,16 @@ def configure_optimizers(self):
...
     
-    def _generate_noise(self, n_tokens):
-        return torch.randn(n_tokens, self.hparams.noise_dim)
+    def _generate_noise(self, batch_size, n_tokens):
+        return torch.randn(batch_size, n_tokens, self.hparams.noise_dim)
     
+    def _update_gumbel_temp(self):
+        progress = self.trainer.global_step / (0.8 * self.trainer.max_epochs * \
+            len(self.hparams.datamodule.train_dataloader()))
+        progress = 1 - (1 - progress)**2
```
Comment (remilvus) [inline hadml/models/cgan/multihadron_event_gan.py:173]: [Bug] `progress = 1 - (1 - max(0, progress))**2`. Otherwise gumbell temperature will start rising near the end of the training

File: hadml/models/cgan/multihadron_event_gan.py [hunk 16]
```diff
@@ -138,11 +256,40 @@ def _prepare_plots(self, predictions=None, truths=None, sentence_stats=None, clu
...
+            if self.hadron_stats is None:
+                with open(self.training_stats_filename, "rb") as f:
+                    stats = np.load(f, allow_pickle=True).item()
+                self.hadron_stats = {
+                    "momentum_mean" : stats["hadron_momentum_mean"], "momentum_std" : stats["hadron_momentum_std"],
+                    "energy_mean" : stats["hadron_energy_mean"], "energy_std" : stats["hadron_energy_std"], 
+                }
+            mean, std = self.hadron_stats["energy_mean"], self.hadron_stats["energy_std"]
+            condition = (preds_kin[:, 0] >= (mean - 3*std)) * (preds_kin[:, 0] <= (mean + 3*std))
+            clipped_preds_energy = preds_kin[condition][:, 0]
```
Comment (remilvus) [inline hadml/models/cgan/multihadron_event_gan.py:275]: It's a good idea to clip the energy/kinematics, but I would avoid dropping results outside of the range: instead clip them, as the variable name suggests: https://pytorch.org/docs/stable/generated/torch.clamp.html#torch.clamp
(or at least rename the variable)

Also, I know that it works, but multiplication `*` is a bit confusing in terms of doing boolean operations (use boolean operations :wink:)

**Review by remilvus — COMMENTED** [review]
Looks good overall. This PR is pretty big, so avoid adding any new changes (other than the gumbell bug) - we can handle them in separate PRs. Also - if anything related to style I mentioned  would be time consuming feel free to ignore it.

**Review by remilvus — APPROVED** [review]

File: hadml/datamodules/gan_datamodule.py (hunk 11 above)
Comment (TairYerniyazov) [reply hadml/datamodules/gan_datamodule.py:418]: Done.

File: hadml/datamodules/components/herwig.py (hunk 12 above)
Comment (TairYerniyazov) [reply hadml/datamodules/components/herwig.py]: Done.

File: hadml/datamodules/components/herwig.py (hunk 13 above)
Comment (TairYerniyazov) [reply hadml/datamodules/components/herwig.py]: Done.

File: hadml/models/cgan/multihadron_event_gan.py (hunk 14 above)
Comment (TairYerniyazov) [reply hadml/models/cgan/multihadron_event_gan.py]: Done.

File: hadml/models/cgan/multihadron_event_gan.py (hunk 15 above)
Comment (TairYerniyazov) [reply hadml/models/cgan/multihadron_event_gan.py]: Fixed.

File: hadml/models/cgan/multihadron_event_gan.py (hunk 16 above)
Comment (TairYerniyazov) [reply hadml/models/cgan/multihadron_event_gan.py]: Thanks! I have changed the `*` operator to `torch.logical_and(...)`. 

As for clipping the four-vectors, I want to limit each of the distribution `(E, px, py, pz)` to accept 99.7% of all the hadrons. That is why I create four conditions which I apply afterwards to filter out the outliers (and it even seems to work as the visualisation looks better). Therefore, clipping (or 'trimming' the original set) is exactly what I want. Using `torch.clamp(...)` would only replace the outliers with the provided min and max values I would then have to hide in the histograms anyway. This would just add extra effort without any real benefit. I renamed the variables, so thanks for that.

**Review by remilvus — APPROVED** [review]
